### PR TITLE
Constraint generation

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by mlutze** [comment]
I've put this on a Flix branch so we can see it and I can merge into it separately.

**Comment by mlutze** [comment]
```
✓ Reader                                    27.2ms
✓ Parser                                   947.1ms
✓ Weeder                                   513.0ms
✓ Namer                                    241.4ms
✓ Resolver                                 631.7ms
✓ Kinder                                   285.8ms
✓ Deriver                                    8.4ms
✓ ConstraintGeneration                     118.6ms
✓ Typer                                   2188.3ms
    ClassEnv                             (   1.4ms)
    EqualityEnv                          (   1.1ms)
    Classes                              ( 556.2ms)
    Instances                            ( 433.6ms)
    Defs                                 (1090.4ms)
    TypeInference                        (2084.9ms)
    Classes                              (  11.2ms)
    Instances                            (  15.6ms)
    Defs                                 (  42.6ms)
    Enums                                (   1.6ms)
    RestrictableEnums                    (   0.2ms)
    Effs                                 (   0.6ms)
    TypeAliases                          (   1.0ms)
    ClassEnv                             (   0.6ms)
    EqualityEnv                          (   1.1ms)
    TypeReconstruction                   (  95.4ms)
✓ EntryPoint                                 0.2ms
✓ Statistics                               116.7ms
✓ Instances                                268.1ms
✓ Stratifier                               105.0ms
    Compute Global Dependency Graph      (  33.7ms)
    Stratify Defs                        (  59.7ms)
    Stratify Instance Defs               (   8.6ms)
✓ Regions                                   39.4ms
✓ PatternExhaustiveness                    110.1ms
✓ Redundancy                               216.6ms
✓ Safety                                    57.1ms
✓ Lowering                                 162.4ms
✓ EarlyTreeShaker                            8.4ms
✓ Monomorph                                  0.8ms
✓ MonomorphEnums                             0.4ms
✓ Simplifier                                 0.4ms
✓ ClosureConv                                0.2ms
✓ LambdaLift                                 0.3ms
✓ Tailrec                                    0.2ms
✓ Optimizer                                  6.1ms
    OccurrenceAnalyzer                   (   1.8ms)
    Inliner                              (   0.3ms)
    OccurrenceAnalyzer                   (   0.2ms)
    Inliner                              (   0.0ms)
✓ LateTreeShaker                             0.7ms
✓ Reducer                                    0.6ms
✓ VarNumbering                               0.3ms
```

**Comment by magnus-madsen** [comment]
Ill take a look as soon as able. Perhaps we can also look through it together tomorrow?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala [hunk 1]
```diff
@@ -0,0 +1,1145 @@
...
+    }
+
+    // return the result
+    result
+  }
+
+  sealed class Constraint
+
+  object Constraint {
+    case class Equality(tpe1: Type, tpe2: Type, lenv: LevelEnv, loc: SourceLocation) extends Constraint
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:69]: I think my most important comment is that we need two Equality constraints: 

- One where types must be equals, and
- One where we expect one type.

So we can keep generating useful error messages.

**Comment by magnus-madsen** [comment]
!!!jar

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala [hunk 2]
```diff
@@ -0,0 +1,1145 @@
...
+    * Exits the type variable's scope in the context.
+    */
+  def exitScopeM(sym: Symbol.KindedTypeVarSym)(implicit c: Context): Unit = {
+    c.lenv = c.lenv.exitScope(sym)
+  }
+
+  /**
+    * Contains information to perform type unification in an expression.
+    */
+  case class Context(constrs: ListBuffer[Constraint], var renv: RigidityEnv, var lenv: LevelEnv)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:178]: Do we want to move Context into its own class together with its unifyType methods so it becomes like its own thing? Or is it fine to have it here?

**Comment by magnus-madsen** [comment]
The printing of constraints is not really readable :)

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala (hunk 1 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:69]: I was thinking that we could keep it with two constraints and then add a field `provenance` or whatever that tells where the constraint came from. That way the constraint reduction doesn't need to care about it.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala (hunk 2 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:178]: Yes, I think we should move it.

**Comment by mlutze** [comment]
> The printing of constraints is not really readable :)

I'll see what I can do.

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
Lets do it. Keep one PR, later break it into smaller pieces.

**Pull request closed without merging**